---
title: "Holothruia atra Anlaysis"
author: "Nikko"
date: "7/11/2021"
output: html_document
---

##Set up

```{r}
library(knitr)
library(rentrez)
library(pegas)
library(tidyverse)
library(plyr)
library(geomedb)
library(strataG)
```
# Query metadata

Just looked through general longitude and latitudes to get majority of the data

```{r}
holothuria_metadata <- queryMetadata(
  entity = "Sample",
  query = "genus = Holothuria AND specificEpithet = atra AND _exists_:sequence",
  select = "Event")
View(holothuria_metadata$Event)
View(holothuria_metadata$Sample)

holothuria_flat <- left_join(holothuria_metadata$Sample, holothuria_metadata$Event, by = "eventID")
```

## Subset the metadata

```{r}
holothuria_event_info <-holothuria_metadata$Event[,c("locality", "country", "decimalLatitude", "decimalLongitude", "eventID")]
kable(holothuria_event_info)
```

#Lumping and cleaning locality names
```{r}
#Lump Indonesia together
holothuria_metadata$Event$locality[which(holothuria_metadata$Event$country == "Indonesia")] <- "Indonesia"
#Uhh idk if that did anything let's just pause
```

# Query FASTA

```{r}
listLoci()
```

```{r}
holothuria_CO1 <- querySanger(query = "genus = Holothuria AND specificEpithet = atra AND _exists_:sequence", 
                              locus = "CO1")
holothuria_CO1
```

##Genetic data
```{r}
#Overwrite the names from the FASTA to have shorter values
names(holothuria_CO1) <- str_replace(string=names(holothuria_CO1),
                                     pattern=".+\\[tissueID = (.+?)\\].+", replacement = "\\1")
```

#Sequence alignment
```{r}
#Align the data to ensure that each site is homologous
holothuria_CO1 <- muscle(holothuria_CO1)
image.DNAbin(holothuria_CO1)
```

150 to 200 looks interesting...
```{r}
image.DNAbin(holothuria_CO1[1:20, 150:200])
```

Let's trim the missing parts of the Skillings sequences

```{r}
image.DNAbin(holothuria_CO1[,60:482])

holothuria_CO1 <- holothuria_CO1[,60:482]

```

##Merging metadata with genetic data
```{r}
#Eric already did a left join earlier
head(holothuria_flat$materialSampleID)

```

```{r}
head(rownames(holothuria_CO1))
```

Put the metadata in the same order as the genetic data.

```{r}
holothuria_flat <- holothuria_flat[order(holothuria_flat$materialSampleID),]
holothuria_CO1 <- holothuria_CO1[order(rownames(holothuria_CO1)),]
```

##Creating a haplotype network
```{r}
#Make a list of populations that were sampled with locality and country names
holothuria_pop <- str_remove_all(holothuria_flat$locality,'\\"')
a <- dist.dna(holothuria_CO1)
b <- pegas::haplotype(holothuria_CO1)
b <- sort(b, what="label")

#Create the network
holothuria_net <- pegas::haploNet(b)

#Make a table of which haplotypes occur in which populations
c <-stack(setNames(attr(b, "index"), rownames(b)))
c <-c[order(c$values),]
holo_ind.hap<-table(hap=c$ind, holothuria_pop=holothuria_pop)

#Skip distinguishing oceans for now?
plot(holothuria_net,size=attr(holothuria_net, "freq"), scale.ratio=0.5, pie=holo_ind.hap, legend=F, labels=F, threshold=0, show.mutation=2, fast=T)
```

##Diversity statistics

#Haplotype diversity

```{r}
#Create a copy of dataset
holothuria_CO1_copy <- holothuria_CO1
rownames(holothuria_CO1_copy) <-holothuria_pop
stratastat<-function(x,holothuria_pop=holothuria_pop, fun=nuc.div){
  stats<-NULL
  for(p in unique(holothuria_pop)){
    stat<-fun(x[grep(p,rownames(x)),])
  }
  
  names(stats)<-unique(holothuria_pop)
  return(stats)
}

holothuria_hapdivs<-stratastat(holothuria_CO1_copy, holothuria_pop =holothuria_pop, fun=hap.div)

holothuria_hapdivs
```
#Nucleotide diversity
```{r}
holothuria_nucdivs<-stratastat(holothuria_CO1_copy, pop = holothuria_pop, fun = nuc.div)
holothuria_nucdivs
```
#Double check calcs
```{r}
holothria_z<-sequence2gtypes(holothuria_CO1, strata=holothuria_pop)
holothria_z<-labelHaplotypes(holothria_z)
holothria_z
```
#Make a table of summary stats
```{r}
samples<-summary(holothria_z$gtypes)$strata.smry[,1]
haplotypes<-summary(holothria_z$gtypes)$strata.smry[,3]
diversities<-cbind("# Samples" = samples, "# Haplotypes" = haplotypes, "Nucleotide Diversity" = holothuria_nucdivs, "Haplotype Diversity" = holothuria_hapdivs)
kable(diversities, digits=3)
```

#Fst table
```{r}
pairwise_phi<-pairwiseTest(holothria_z$gtypes,stats="phist",nrep=1000,quietly=T,model="raw")

#PhiST in lower triangle, p-value in upper triangle
kable(pairwise_phi$pair.mat$PHIst, digits=3)
```
```{r}
phiST<-pairwise_phi$pair.mat$PHIst
phiST[upper.tri(phiST)] = t(phiST)[upper.tri(phiST)]
heatmap(phiST)
```

